internal/xresource/unstructured/xreconcile/reconciler.go

```diff
@@ -0,0 +1,139 @@
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	"github.com/crossplane/crossplane/internal/xresource/unstructured/xreconcile/xlogging"
+)
+
+const (
+	reconcileTimeout = 1 * time.Minute
+
+	// ErrClientGet is the error message for client.Get() failures.
+	ErrClientGet = "cannot get %s"
+	// ErrClientStatusUpdate is the error message for client.Status().Update() failures.
+	ErrClientStatusUpdate = "cannot update %s status"
+)
+
+// A ReconcilerOption configures a Reconciler.
+type ReconcilerOption[object client.Object] func(*objectReconcilerAdapter[object])
+
+// WithReconcileTimeout specifies the timeout for each reconciliation loop.
+func WithReconcileTimeout[object client.Object](timeout time.Duration) ReconcilerOption[object] {
+	return func(r *objectReconcilerAdapter[object]) {
+		r.timeout = timeout
+	}
+}
+
+// WithLogger specifies how the Reconciler should log messages.
+func WithLogger[object client.Object](l logging.Logger) ReconcilerOption[object] {
+	return func(r *objectReconcilerAdapter[object]) {
+		r.log = l
+	}
+}
+
+// AsUnstructuredReconciler returns a reconcile.Reconciler implementation that wraps unstructured reconcilers that would
+// like to adopt the reconcile.AsReconciler pattern but for reconcilers that do not conform to a concreate type.
+func AsUnstructuredReconciler[object client.Object](client client.Client, rec reconcile.ObjectReconciler[object], constructorFn func() object, o ...ReconcilerOption[object]) reconcile.Reconciler {
+	r := &objectReconcilerAdapter[object]{
+		objReconciler: rec,
+		client:        client,
+		constructorFn: constructorFn,
+		timeout:       reconcileTimeout,
+		log:           logging.NewNopLogger(),
+	}
+
+	for _, ro := range o {
+		ro(r)
+	}
+
+	return r
+}
+
+type objectReconcilerAdapter[object client.Object] struct {
+	objReconciler     reconcile.ObjectReconciler[object]
+	client            client.Client
+	constructorFn     func() object
+	timeout           time.Duration
+	log               logging.Logger
+	skipStatusUpdates bool
+}
+
+// Reconcile implements Reconciler.
+func (r *objectReconcilerAdapter[object]) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("request", req)
+	log.Debug("Reconciling")
+
+	// Preamble
+
+	rctx, cancel := context.WithTimeout(ctx, r.timeout)
+	defer cancel()
+
+	o := r.constructorFn()
+	if err := r.client.Get(rctx, req.NamespacedName, o); err != nil {
+		msg := fmt.Sprintf(ErrClientGet, o.GetObjectKind().GroupVersionKind().Kind)
+		log.Debug(msg, "error", err)
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), msg)
+	}
+	original := o.DeepCopyObject()
+
+	log = log.WithValues(
+		"uid", o.GetUID(),
+		"version", o.GetResourceVersion(),
+	)
+	if extName := meta.GetExternalName(o); extName != "" {
+		log = log.WithValues("external-name", extName)
+	}
+	rctx = xlogging.WithLogger(rctx, log)
```

Comment:
I still don't feel good about passing loggers in the context. I'm happy to adopt the wrapper pattern this PR proposes, especially in light of https://github.com/crossplane/crossplane/issues/6571, but this is a blocker. I'd like to move forward with this PR without the logging changes at least for now - we can debate them more later.

Reply:
ok I can do that!